File: practice/src/main/kotlin/ar/com/miura/jvm/backtracing/WordSearch.java
package ar.com.miura.jvm.backtracing;

public class WordSearch {

    public boolean exist(char[][] board, String word) {

        int rowQty = board.length;
        int colQty = board[0].length;

        boolean[][] visited = new boolean[rowQty][colQty];

        for(int i=0;i<rowQty;i++) {
            for(int j=0;j<colQty;j++) {
                if (word.charAt(0)==board[i][j] && search(board, i, j, 0, visited, word)) {
                    return true;
                }
            }
        }

        return false;
    }

    /**

     **/
    private boolean search(char[][] board, int row, int col, int index, boolean[][] visited, String word) {

        int rowQty = board.length;
        int colQty = board[0].length;

        /*
        End of backtracking
        */
        if (index==word.length()) {
            return true;
        }


        /*
        Check out of board
        */
        if (row<0 || row>=rowQty || col < 0 || col >= colQty) {
            return false;

            /**
             Not a valid next word
             **/
        } else if (board[row][col]!=word.charAt(index)) {
            return false;

            /**
             Avoid visited col and rows
             **/
        } else if (visited[row][col]) {
            return false;

        }

        /*
        Visit the position
        check
            * up
            * right
            * down
            * left
        */
        visited[row][col] = true;
        if ( (search(board, row-1, col, index+1, visited, word)) ||
                (search(board, row, col+1, index+1, visited, word)) ||
                (search(board, row+1, col, index+1, visited, word)) ||
                (search(board, row, col-1, index+1, visited, word)) ) {

            return true;

        }

        visited[row][col] = false;

        return false;

    }
}

File: practice/src/main/kotlin/ar/com/miura/lc/backtracking/WordSearch.kt
package ar.com.miura.lc.backtracking

/**
 * https://leetcode.com/explore/featured/card/top-interview-questions-medium/109/backtracking/797/
 */
class WordSearch {
    fun exist(board: Array<CharArray>, word: String): Boolean {

        var rows = board.size
        var cols = board[0].size

        for (row in 0..(rows - 1)) {
            for (col in 0..(cols - 1)) {
                var firstChar = board[row][col]
                if (firstChar == word.get(0)) {
                    var find = backtrack(board, row, col, word, 0)
                    if (find) {
                        return true
                    }
                }
            }
        }
        return false

    }

    private fun backtrack(board: Array<CharArray>, row: Int, col: Int, word: String, index: Int): Boolean {

        if (index >= word.length) {
            return true
        }

        var rows = board.size
        var cols = board[0].size

        if (row < 0 || row >= rows || col < 0 || col >= cols) {
            return false
        }

        if (board[row][col] != word.get(index)) {
            return false
        }

        var originalChar = board[row][col]
        //mark as visited
        board[row][col] = '#'

        var movements = listOf<IntArray>(
            intArrayOf(1, 0),    //up
            intArrayOf(-1, 0),   //down
            intArrayOf(0, -1),   //left
            intArrayOf(0, 1)     //right
        )

        var find = false
        for (movement in movements) {

            var newI = row + movement[0]
            var newJ = col + movement[1]

            find = backtrack(board, newI, newJ, word, index + 1)
            if (find) {
                break
            }

        }
        board[row][col] = originalChar
        return find


    }
}